src/coprocessor/dag/batch/executors/limit.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use tipb::expression::FieldType;
+
+use crate::coprocessor::dag::batch::interface::*;
+use crate::coprocessor::Result;
+
+/// Executor that retrieves rows from the source executor
+/// and only produces part of the rows.
+pub struct BatchLimitExecutor<Src: BatchExecutor, C: ExecSummaryCollector> {
+    src: Src,
+    remaining_rows: usize,
+    summary_collector: C,
+}
+
+impl<Src: BatchExecutor, C: ExecSummaryCollector> BatchLimitExecutor<Src, C> {
+    pub fn new(src: Src, limit: usize, summary_collector: C) -> Result<Self> {
+        if limit == 0 {
+            return Err(box_err!("limit should not be zero"));
+        }
+        Ok(Self {
+            src,
+            remaining_rows: limit,
+            summary_collector,
+        })
+    }
+}
+
+impl<Src: BatchExecutor, C: ExecSummaryCollector> BatchExecutor for BatchLimitExecutor<Src, C> {
+    #[inline]
+    fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+        let timer = self.summary_collector.start_record_duration();
+        self.summary_collector.inc_iterations();
+
+        let mut result = self.src.next_batch(expect_rows);
+        if result.data.rows_len() < self.remaining_rows {
+            self.remaining_rows -= result.data.rows_len();
+        } else {
+            result.data.truncate(self.remaining_rows);
+            self.remaining_rows = 0;
+            result.is_drained = Ok(true);
+        }
+        self.summary_collector
+            .inc_produced_rows(result.data.rows_len());
+        self.summary_collector.inc_elapsed_duration(timer);
+        result
+    }
+
+    #[inline]
+    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
+        self.src.collect_statistics(destination);
+        self.summary_collector
+            .collect_into(&mut destination.summary_per_executor)
+    }
+
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.src.schema()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+    use crate::coprocessor::codec::data_type::VectorValue;
+    use crate::coprocessor::dag::batch::statistics::{
+        ExecSummaryCollectorDisabled, ExecSummaryCollectorEnabled,
+    };
+    use crate::coprocessor::dag::expr::EvalConfig;
+    use cop_datatype::{EvalType, FieldTypeAccessor, FieldTypeTp};
+    use tipb::expression::FieldType;
+
+    struct MockExecutor {
```

Comment:
I feel like this predefined and fixed schema and data in MockExecutor makes these tests more complicated: you need to add more and more flags here in order to control different behaviour (like `zero_rows_for_next_batch`) required by each test case, which just breaks the DAMP rule. A more preferred way should be writing the behaviour just inside the corresponding test case. To do this, you can make this MockExecutor do less things, instead of a universal MockExecutor that try to do everything.

Reply:
I think the unit test should focus on current change, so I'd like to remove the tests related to getting 0 rows in `next_batch` since we already have tested about get rows less than `limit`